include/experimental/kernel/col_major_shuf/col_major_shuf_xe.hpp

```diff
@@ -0,0 +1,152 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+/// @file
+/// C++ API
+
+#pragma once
+
+#include <experimental/kernel/col_major_shuf/api.hpp>
+#include <experimental/kernel/col_major_shuf/common.hpp>
+#include <experimental/kernel/col_major_shuf/config.hpp>
+
+namespace gpu::xetla::kernel {
+template <
+    typename dtype_in_,
+    typename dtype_out_,
+    typename dtype_gidx_,
+    mem_layout mem_layout_in_,
+    typename col_major_shuf_attr_,
+    gpu_arch arch_>
+struct col_major_shuf_t<
+    dtype_in_,
+    dtype_out_,
+    dtype_gidx_,
+    mem_layout_in_,
+    col_major_shuf_attr_,
+    arch_,
+    std::enable_if_t<arch_ == gpu_arch::XeHpg || arch_ == gpu_arch::XeLpg>> {
+  using dtype_in = dtype_in_;
+  using dtype_out = dtype_out_;
+  using dtype_gidx = dtype_gidx_;
+  using col_major_shuf_attr = col_major_shuf_attr_;
+
+  static constexpr mem_layout mem_layout_in = mem_layout_in_;
+
+  static_assert(
+      std::is_same<dtype_in, dtype_out>::value,
+      "only support in/out data type must be same now.");
+  static_assert(
+      mem_layout_in == mem_layout::row_major,
+      "only support row_major input now.");
+  static_assert(
+      std::is_same<dtype_gidx, uint32_t>::value,
+      "dtype_gidx must be uint32_t");
+
+  static constexpr uint32_t wg_tile_x = col_major_shuf_attr::wg_tile_x;
+  static constexpr uint32_t wg_tile_y = col_major_shuf_attr::wg_tile_y;
+  static constexpr uint32_t sg_tile_x = col_major_shuf_attr::sg_tile_x;
+  static constexpr uint32_t sg_tile_y = col_major_shuf_attr::sg_tile_y;
+
+  static constexpr uint32_t tile_size_x = sg_tile_x;
+  static constexpr uint32_t tile_size_y = sg_tile_y;
+
+  static constexpr uint32_t block_size_x =
+      col_major_shuf_attr::load_block_size; // TODO(zhe:) add load block size
+                                            // check under different arch
+
+  static constexpr uint32_t dev_mem_align = 64;
+  using mem_desc_store_tile_t = mem_desc_t<
+      dtype_in,
+      mem_layout_in,
+      mem_space::global,
+      dev_mem_align / sizeof(dtype_in)>;
+  using store_tile_desc_t = subgroup::tile_desc_t<
+      tile_size_x,
+      tile_size_y,
+      block_size_x,
+      tile_size_y,
+      reg_layout::tiled>;
+  using store_tile_t = subgroup::tile_t<dtype_out, store_tile_desc_t>;
+  using store_tile_payload_t = subgroup::mem_payload_t<
+      mem_desc_store_tile_t,
+      store_tile_desc_t,
+      msg_type::block_2d,
```

Comment:
```suggestion
      subgroup::msg_type_v<store_tile_desc_t, dtype_out>,
```

Reply:
this will lead store_tile_payload can't obtain base_y field which needed in gather instruction.
i also did some tests and looks like it will not affect perf too much when we always use block_2d to load/store in shuffle kernel.
anyway,  this is a reasonable suggestion, I will try to use block_1d msg_type when m==1 if there have a chance.